File: library/kotlin/src/io/envoyproxy/envoymobile/ResponseHeaders.kt
package io.envoyproxy.envoymobile

/*
 * Headers representing an inbound response.
 */
class ResponseHeaders : Headers {
  /**
   * Internal constructor used by builders.
   *
   * @param headers: Headers to set.
   */
  internal constructor(headers: Map<String, List<String>>) : super(headers)

  /**
   * HTTP status code received with the response.
   */
  val httpStatus: Int? by lazy { value(":status")?.first()?.toIntOrNull() }

  /**
   * Convert the headers back to a builder for mutation.
   *
   * @return ResponseHeadersBuilder, The new builder.
   */
  fun toResponseHeadersBuilder(): ResponseHeadersBuilder {
    return ResponseHeadersBuilder(this.headers.mapValues { it.value.toMutableList() }
      .toMutableMap())
  }
}

File: library/swift/src/ResponseHeaders.swift
import Foundation

/// Headers representing an inbound response.
@objcMembers
public final class ResponseHeaders: Headers {
  /// HTTP status code received with the response.
  /// TODO: in Kotlin this blows up if the value of status cannot be translated to an int.
  public var httpStatus: Int? {
    return self.value(forName: ":status")?.first.flatMap(Int.init)
  }
}
